x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/index/IndexResolver.java

```diff
@@ -394,108 +379,127 @@ private static EsField createField(String fieldName, String typeName, Map<String
+        Map<String, Fields> indices = new LinkedHashMap<>(resolvedIndices.size());
+        Pattern pattern = javaRegex != null ? Pattern.compile(javaRegex) : null;
+
+        // sort fields in reverse order to build the field hierarchy
+        Set<Entry<String, Map<String, FieldCapabilities>>> sortedFields = new TreeSet<>(
+                Collections.reverseOrder(Comparator.comparing(Entry::getKey)));
+
+        sortedFields.addAll(fieldCaps.entrySet());
 
-        // Make sure that the index contains only a single type
-        MappingMetaData singleType = null;
-        List<String> typeNames = null;
-        for (ObjectObjectCursor<String, MappingMetaData> type : mappings) {
-            //Default mappings are ignored as they are applied to each type. Each type alone holds all of its fields.
-            if ("_default_".equals(type.key)) {
+        for (Entry<String, Map<String, FieldCapabilities>> entry : sortedFields) {
+            String fieldName = entry.getKey();
+            Map<String, FieldCapabilities> types = entry.getValue();
+            
+            // ignore size added by the mapper plugin
+            if (FIELD_NAMES_BLACKLIST.contains(fieldName)) {
                 continue;
             }
-            if (singleType != null) {
-                // There are more than one types
-                if (typeNames == null) {
-                    typeNames = new ArrayList<>();
-                    typeNames.add(singleType.type());
+
+            // apply verification
+            final InvalidMappedField invalidField = validityVerifier.apply(fieldName, types);
+            
+            // filter meta fields and unmapped
+            FieldCapabilities unmapped = types.get(UNMAPPED);
+            List<String> unmappedIndices = unmapped != null ? asList(unmapped.indices()) : emptyList();
+
+            // check each type
+            for (Entry<String, FieldCapabilities> typeEntry : types.entrySet()) {
+                FieldCapabilities typeCap = typeEntry.getValue();
+                String[] capIndices = typeCap.indices();
+                
+                // Skip internal fields (name starting with underscore and its type reported by field_caps starts
+                // with underscore as well). A meta field named "_version", for example, has the type named "_version".
+                if (typeEntry.getKey().startsWith("_") && typeCap.getType().startsWith("_")) {
+                    continue;
+                }
+
+                // compute the actual indices - if any are specified, take into account the unmapped indices
+                List<String> concreteIndices = null;
+                if (capIndices != null) {
+                    // make a copy of the list to be able to modify it
+                    concreteIndices = new ArrayList<>(asList(capIndices));
+                    if (unmappedIndices.isEmpty() == false) {
+                        concreteIndices.removeAll(unmappedIndices);
```

Comment:
Maybe not so important: I would use a `Set` for the `unmappedIndices` and use `contains()` to speed up the removal. This way you can iterate on the String[] and create the `concreteIndices` list.